### ISSUE TITLE
[pbr] issue: `Installing fw4 nft file` fails with straightforward rule?

### ISSUE THREAD

**Issue by posita** [stangri/openwrt_packages#204]
**Describe the bug**

**Your configs**

1. ```/etc/config/dhcp```

```sh

config dnsmasq
	option domainneeded '1'
	option boguspriv '1'
	option filterwin2k '0'
	option localise_queries '1'
	option rebind_protection '1'
	option rebind_localhost '1'
	option local '/lan/'
	option domain '...'
	option expandhosts '1'
	option nonegcache '0'
	option cachesize '1000'
	option authoritative '1'
	option readethers '1'
	option leasefile '/tmp/dhcp.leases'
	option resolvfile '/tmp/resolv.conf.d/resolv.conf.auto'
	option nonwildcard '1'
	option localservice '1'
	option ednspacket_max '1232'
	option filter_aaaa '0'
	option filter_a '0'
	option sequential_ip '1'
	option confdir '/tmp/dnsmasq.d'

config dhcp 'lan'
	option interface 'lan'
	option start '258'
	option limit '125'
	option leasetime '12h'
	option dhcpv4 'server'

config dhcp 'wan'
	option interface 'wan'
	option ignore '1'

config odhcpd 'odhcpd'
	option maindhcp '0'
	option leasefile '/tmp/hosts/odhcpd'
	option leasetrigger '/usr/sbin/odhcpd-update'
	option loglevel '4'

config dhcp 'iot'
	option interface 'iot'
	option start '258'
	option limit '125'
	option leasetime '12h'

config dhcp 'guest'
	option interface 'guest'
	option start '258'
	option limit '125'
	option leasetime '12h'
	list dhcp_option '6,9.9.9.9'
	list dhcp_option '6,149.112.112.112'
	option force '1'

```

2. ```/etc/config/network```

```sh

config interface 'loopback'
	option device 'lo'
	option proto 'static'
	option ipaddr '127.0.0.1'
	option netmask '255.0.0.0'

config globals 'globals'

config device
	option name 'br-lan'
	option type 'bridge'
	list ports 'lan2'
	list ports 'lan3'
	list ports 'lan4'
	list ports 'lan5'

config device
	option name 'lan1'
	option macaddr 'xx:xx:xx:xx:xx:xx'

config device
	option name 'lan2'
	option macaddr 'xx:xx:xx:xx:xx:xx'

config device
	option name 'lan3'
	option macaddr 'xx:xx:xx:xx:xx:xx'

config device
	option name 'lan4'
	option macaddr 'xx:xx:xx:xx:xx:xx'

config device
	option name 'lan5'
	option macaddr 'xx:xx:xx:xx:xx:xx'

config interface 'lan'
	option device 'br-lan.32'
	option proto 'static'
	option ipaddr '192.168.32.1'
	option netmask '255.255.254.0'
	option ipv6 '0'
	option delegate '0'

config device
	option name 'eth1'
	option macaddr 'xx:xx:xx:xx:xx:xx'

config interface 'wan'
	option device 'eth1'
	option proto 'dhcp'
	option hostname '*'
	option ipv6 '0'
	option delegate '0'
	option peerdns '0'
	list dns '9.9.9.10'
	list dns '149.112.112.10'

config interface 'parachute'
	option proto 'static'
	option device 'lan1'
	option ipaddr '10.10.10.10'
	option netmask '255.255.255.0'
	option defaultroute '0'
	option delegate '0'
	option ipv6 '0'

config interface 'iot'
	option device 'br-lan.34'
	option proto 'static'
	option ipv6 '0'
	option delegate '0'
	option ipaddr '192.168.34.1'
	option netmask '255.255.254.0'

config interface 'guest'
	option device 'br-lan.64'
	option proto 'static'
	option ipv6 '0'
	option delegate '0'
	option ipaddr '192.168.64.1'
	option netmask '255.255.254.0'

config interface 'vpn'
	option proto 'none'
	option auto '0'
	option device 'ovpnc0'
	option delegate '0'
	option ipv6 '0'

config bridge-vlan
	option device 'br-lan'
	option vlan '32'
	list ports 'lan2:u*'
	list ports 'lan5:t'

config bridge-vlan
	option device 'br-lan'
	option vlan '34'
	list ports 'lan3:u*'
	list ports 'lan5:t'

config bridge-vlan
	option device 'br-lan'
	option vlan '64'
	list ports 'lan5:t'

```

3. ```/etc/config/firewall```

```sh


config defaults
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	option synflood_protect '1'

config zone
	option name 'lan'
	list network 'lan'
	list network 'parachute'
	option input 'ACCEPT'
	option output 'ACCEPT'
	option forward 'REJECT'

config zone
	option name 'wan'
	list network 'wan'
	list network 'vpn'
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	option masq '1'
	option mtu_fix '1'

config forwarding
	option src 'lan'
	option dest 'wan'

config rule
	option name 'Allow-DHCP-Renew'
	option src 'wan'
	option proto 'udp'
	option dest_port '68'
	option target 'ACCEPT'
	option family 'ipv4'

config rule
	option name 'Allow-Ping'
	option src 'wan'
	option proto 'icmp'
	option icmp_type 'echo-request'
	option family 'ipv4'
	option target 'ACCEPT'

config rule
	option name 'Allow-IGMP'
	option src 'wan'
	option proto 'igmp'
	option family 'ipv4'
	option target 'ACCEPT'

config rule
	option name 'Allow-DHCPv6'
	option src 'wan'
	option proto 'udp'
	option dest_port '546'
	option family 'ipv6'
	option target 'REJECT'

config rule
	option name 'Allow-MLD'
	option src 'wan'
	option proto 'icmp'
	option src_ip 'fe80::/10'
	list icmp_type '130/0'
	list icmp_type '131/0'
	list icmp_type '132/0'
	list icmp_type '143/0'
	option family 'ipv6'
	option target 'REJECT'

config rule
	option name 'Allow-ICMPv6-Input'
	option src 'wan'
	option proto 'icmp'
	list icmp_type 'echo-request'
	list icmp_type 'echo-reply'
	list icmp_type 'destination-unreachable'
	list icmp_type 'packet-too-big'
	list icmp_type 'time-exceeded'
	list icmp_type 'bad-header'
	list icmp_type 'unknown-header-type'
	list icmp_type 'router-solicitation'
	list icmp_type 'neighbour-solicitation'
	list icmp_type 'router-advertisement'
	list icmp_type 'neighbour-advertisement'
	option limit '1000/sec'
	option family 'ipv6'
	option target 'REJECT'

config rule
	option name 'Allow-ICMPv6-Forward'
	option src 'wan'
	option dest '*'
	option proto 'icmp'
	list icmp_type 'echo-request'
	list icmp_type 'echo-reply'
	list icmp_type 'destination-unreachable'
	list icmp_type 'packet-too-big'
	list icmp_type 'time-exceeded'
	list icmp_type 'bad-header'
	list icmp_type 'unknown-header-type'
	option limit '1000/sec'
	option family 'ipv6'
	option target 'REJECT'

config rule
	option name 'Allow-IPSec-ESP'
	option src 'wan'
	option dest 'lan'
	option proto 'esp'
	option target 'ACCEPT'

config rule
	option name 'Allow-ISAKMP'
	option src 'wan'
	option dest 'lan'
	option dest_port '500'
	option proto 'udp'
	option target 'ACCEPT'

config zone
	option name 'iot'
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	list network 'iot'

config zone
	option name 'guest'
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	list network 'guest'

config forwarding
	option src 'lan'
	option dest 'iot'

config rule
	option name 'allow-dhcp-iot'
	option src 'iot'
	option dest_port '67'
	option proto 'udp'
	option family 'ipv4'
	option target 'ACCEPT'

config rule
	option name 'allow-dhcp-guest'
	option src 'guest'
	option dest_port '67'
	option proto 'udp'
	option family 'ipv4'
	option target 'ACCEPT'

config zone
	option name 'vpn'
	option network 'vpn'
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	option masq '1'
	option mtu_fix '1'

config forwarding
	option src 'guest'
	option dest 'vpn'

config include 'pbr'
	option fw4_compatible '1'
	option type 'script'
	option path '/usr/share/pbr/firewall.include'

```

4. ```/etc/config/pbr```

```sh

config pbr 'config'
	option enabled '1'
	option verbosity '2'
	option strict_enforcement '1'
	option resolver_set 'dnsmasq.ipset'
	list resolver_instance '*'
	option ipv6_enabled '0'
	list ignored_interface 'vpnserver'
	option boot_timeout '30'
	option rule_create_option 'add'
	option procd_boot_delay '0'
	option procd_reload_delay '6'
	option webui_show_ignore_target '0'
	option nft_rule_counter '0'
	option nft_set_auto_merge '1'
	option nft_set_counter '0'
	option nft_set_flags_interval '1'
	option nft_set_flags_timeout '0'
	option nft_set_policy 'performance'
	list webui_supported_protocol 'all'
	list webui_supported_protocol 'tcp'
	list webui_supported_protocol 'udp'
	list webui_supported_protocol 'tcp udp'
	list webui_supported_protocol 'icmp'
	option secure_reload '1'
	list supported_interface 'vpn'

config include
	option path '/usr/share/pbr/pbr.user.aws'
	option enabled '0'

config include
	option path '/usr/share/pbr/pbr.user.netflix'
	option enabled '0'

config include
	option path '/usr/share/pbr/pbr.user.wg_server_and_client'
	option enabled '0'

config dns_policy
	option name 'Redirect Local IP DNS'
	option src_addr '192.168.1.5'
	option dest_dns '1.1.1.1'
	option enabled '0'

config policy
	option name 'Ignore Local Requests'
	option interface 'ignore'
	option dest_addr '10.0.0.0/24 10.0.1.0/24 192.168.100.0/24 192.168.1.0/24'
	option enabled '0'

config policy
	option name 'Plex/Emby Local Server'
	option interface 'wan'
	option src_port '8096 8920 32400'
	option enabled '0'

config policy
	option name 'Plex/Emby Remote Servers'
	option interface 'wan'
	option dest_addr 'plex.tv my.plexapp.com emby.media app.emby.media tv.emby.media'
	option enabled '0'

config policy
	option name 'lan-wan'
	option src_addr '192.168.32.0/24'
	option interface 'wan'
	option dest_addr '192.168.100.1 ticketmaster.com'

```

**To Reproduce**

Context:

```
root@OpenWrt:~# opkg list-installed | grep pbr
luci-app-pbr - 1.1.6
pbr - 1.1.6-22
root@OpenWrt:~# cat /etc/openwrt_release
DISTRIB_ID='OpenWrt'
DISTRIB_RELEASE='23.05.5'
DISTRIB_REVISION='r24106-10cc5fcd00'
DISTRIB_TARGET='mediatek/filogic'
DISTRIB_ARCH='aarch64_cortex-a53'
DISTRIB_DESCRIPTION='OpenWrt 23.05.5 r24106-10cc5fcd00'
DISTRIB_TAINTS=''
root@OpenWrt:~# /etc/init.d/pbr stop
Removing routing for 'wan/eth1/XXX.XXX.XXX.XXX' [✓]
Removing routing for 'vpn/ovpnc0/YYY.YYY.YYY.YYY' [✓]
pbr 1.1.6-22 (nft mode) stopped [✓]
```

1. Create the following policy:

```
uci add pbr policy
uci set 'pbr.@policy[-1].name=lan-wan'
uci set 'pbr.@policy[-1].src_addr=192.168.32.0/23'
uci set 'pbr.@policy[-1].interface=wan'
uci set 'pbr.@policy[-1].dest_addr=192.168.100.1 ticketmaster.com'
uci commit pbr
```

2. Restart pbr and note the error:

```
root@OpenWrt:~# /etc/init.d/pbr start
Using wan interface (on_start): wan
Found wan gateway (on_start): XXX.XXX.XXX.XXX
Setting up routing for 'wan/eth1/XXX.XXX.XXX.XXX' [✓]
Setting up routing for 'vpn/ovpnc0/YYY.YYY.YYY.YYY' [✓]
Routing 'lan-wan' via wan [✓]
Installing fw4 nft file [✗]
pbr 1.1.6-22 monitoring interfaces: wan vpn
pbr 1.1.6-22 (nft mode) started with gateways:
wan/eth1/XXX.XXX.XXX.XXX
vpn/ovpnc0/YYY.YYY.YYY.YYY [✓]
ERROR: Failed to install fw4 nft file '/var/run/pbr.nft'!
```

3. Check the resulting `/var/run/pbr.nft` file:

```
root@OpenWrt:~# nft -c -f /var/run/pbr.nft
/var/run/pbr.nft:9:72-98: Error: No such file or directory
add rule inet fw4 pbr_prerouting ip saddr { 192.168.32.0/23 } ip daddr @pbr_wan_4_dst_ip_cfg096ff5  goto pbr_mark_0x010000 comment "lan-wan"
                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
root@OpenWrt:~# cat /var/run/pbr.nft
#!/usr/sbin/nft -f

add chain inet fw4 pbr_mark_0x010000
add rule inet fw4 pbr_mark_0x010000  mark set mark and 0xff00ffff xor 0x010000
add rule inet fw4 pbr_mark_0x010000 return
add chain inet fw4 pbr_mark_0x020000
add rule inet fw4 pbr_mark_0x020000  mark set mark and 0xff00ffff xor 0x020000
add rule inet fw4 pbr_mark_0x020000 return
add rule inet fw4 pbr_prerouting ip saddr { 192.168.32.0/23 } ip daddr @pbr_wan_4_dst_ip_cfg096ff5  goto pbr_mark_0x010000 comment "lan-wan"
add rule inet fw4 pbr_prerouting ip saddr { 192.168.32.0/23 } ip daddr { 192.168.100.1 }  goto pbr_mark_0x010000 comment "lan-wan"
```

4. Remove the offending rule and restart pbr to remove the error:

```
root@OpenWrt:~# uci del pbr.@policy[-1]
root@OpenWrt:~# uci commit
root@OpenWrt:~# /etc/init.d/pbr restart
Removing routing for 'wan/eth1/XXX.XXX.XXX.XXX' [✓]
Removing routing for 'vpn/ovpnc0/YYY.YYY.YYY.YYY' [✓]
pbr 1.1.6-22 (nft mode) stopped [✓]
Using wan interface (on_start): wan
Found wan gateway (on_start): XXX.XXX.XXX.XXX
Setting up routing for 'wan/eth1/XXX.XXX.XXX.XXX' [✓]
Setting up routing for 'vpn/ovpnc0/YYY.YYY.YYY.YYY' [✓]
Installing fw4 nft file [✓]
pbr 1.1.6-22 monitoring interfaces: wan vpn
pbr 1.1.6-22 (fw4 nft file mode) started with gateways:
wan/eth1/XXX.XXX.XXX.XXX
vpn/ovpnc0/YYY.YYY.YYY.YYY [✓]
```
```
root@OpenWrt:~# nft -c -f /var/run/pbr.nft
[no output]
```

**Expected behavior**
I don't see anything fundamentally wrong with the rule, so I'm confused about the error.

**Policy Routing run-time information**

1. Output of ```/etc/init.d/pbr reload``` with ```verbosity``` set to 2:

```sh
Using wan interface (on_start): wan
Found wan gateway (on_start): XXX.XXX.XXX.XXX
Setting up routing for 'wan/eth1/XXX.XXX.XXX.XXX' [✓]
Setting up routing for 'vpn/ovpnc0/YYY.YYY.YYY.YYY' [✓]
Routing 'lan-wan' via wan [✓]
Installing fw4 nft file [✗]
pbr 1.1.6-22 monitoring interfaces: wan vpn
pbr 1.1.6-22 (nft mode) started with gateways:
wan/eth1/XXX.XXX.XXX.XXX
vpn/ovpnc0/10.96.0.7 [✓]
ERROR: Failed to install fw4 nft file '/var/run/pbr.nft'!
```

4. Output of ```/etc/init.d/pbr status```:

```sh

pbr - environment
pbr 1.1.6-22 running on OpenWrt 23.05.5.

Dnsmasq version 2.90  Copyright (c) 2000-2024 Simon Kelley
Compile time options: IPv6 GNU-getopt no-DBus UBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP conntrack no-ipset nftset auth cryptohash DNSSEC no-ID loop-detect inotify dumpfile

pbr chains - policies
        chain pbr_forward { # handle 59
        }
        chain pbr_input { # handle 60
        }
        chain pbr_output { # handle 61
        }
        chain pbr_postrouting { # handle 63
        }
        chain pbr_prerouting { # handle 62
        }
        chain pbr_dstnat { # handle 58
        }

pbr chains - marking

pbr nft sets

IPv4 table 256 route: default via XXX.XXX.XXX.XXX dev eth1
IPv4 table 256 rule(s):
30000:  from all fwmark 0x10000/0xff0000 lookup pbr_wan
IPv4 table 257 route: default via YYY.YYY.YYY.YYY dev ovpnc0
IPv4 table 257 rule(s):
29998:  from all fwmark 0x20000/0xff0000 lookup pbr_vpn
```

**Comment by posita** [comment]
This problem appears to go away with 1.1.8, but after upgrading (and restarting my configuration from scratch), PBR no longer appears to work reliably. I will file another issue for that problem. (I don't know if this is related to #193.)

**Comment by posita** [comment]
> after upgrading (and restarting my configuration from scratch), PBR no longer appears to work reliably

Following up, I got hit by this: https://forum.openwrt.org/t/pbr-not-working-with-some-domains/183426

So `ifconfig.me` (and quite possibly `ifcfg.me`) don't appear to work with PBR?

**Comment by stangri** [comment]
I'm struggling to understand why validation script did not complain about:
```
	option resolver_set 'dnsmasq.ipset'
```
for the nft-capable version, but I suspect the error comes from the fact that no nft set was created due to incompatible dnsmasq instruction.

> So `ifconfig.me` (and quite possibly `ifcfg.me`) don't appear to work with PBR?

Most likely the user error. Please peruse the README.

**Comment by posita** [comment]
> I'm struggling to understand why validation script did not complain about:
> 
> ```
> 	option resolver_set 'dnsmasq.ipset'
> ```
> 
> for the nft-capable version, but I suspect the error comes from the fact that no nft set was created due to incompatible dnsmasq instruction.

Apologies for not being clear. I omitted that line after upgrading to 1.1.8, so perhaps that would have fixed things for 1.1.6 as well? Not sure, but I'm not sure it matters, since it's resolved. :+1:

**Comment by posita** [comment]
> > So `ifconfig.me` (and quite possibly `ifcfg.me`) don't appear to work with PBR?
> 
> Most likely the user error. Please peruse the README.

~I'm the second person to hit that, and https://forum.openwrt.org/t/pbr-not-working-with-some-domains/183426 doesn't indicate any fix. With respect to the README, which part do you think would be helpful? It's not clear to me why that should silently not work when other domains in the same entry do. Also, those docs appear to be for 1.1.6, not 1.1.8? It's not clear. Can you point me to the right place? Thanks!~ _EDIT: Never mind. It started working all of a sudden. Not sure what introduced a delay for that domain? Working now though. Thanks! :+1:_
